proposals/low-level-struct-improvements.md

```diff
@@ -939,17 +968,135 @@ This will work but is likely going to result in unnecessary code generation. One
 1. `Unsafe.AsRef<T>(in T value)` could expand its existing purpose by changing to `scoped in T value`. This would allow it to both remove `in` and `scoped` from parameters. It then becomes the universal "remove ref safety" method
 2. Introduce a new method whose entire purpose is to remove `scoped`: `ref T Unsafe.AsUnscoped<T>(scoped in T value)`. This removes `in` as well because if it did not then callers still need a combination of method calls to "remove ref safety" at which point the existing solution is likely sufficient.
 
-### What will make C# 11.0
+### Unscoped this by default?
+The design only has two locations which are `scoped` by default: 
+
+- `this` is `scoped ref` 
+- `out` is `scoped ref`
+
+The decision on `out` is to significantly reduce the compat burden of `ref` fields and at the same time is a more natural default. It lets developers actually think of `out` as data flowing outward only where as if it's `ref` then the rules must consider data flowing in both directions. This leads to significant developer confusion.
+
+The decision on `this` this though is undesirable because it means a `struct` cannot return a field by `ref`. This is an important scenario to high perf developers and the `unscoped` keyword was added effectively for this one scenario.
```

Comment:
While it is true that unscoped was primarily meant for this scenario, should this also mention that there are other valid cases where it would also be useful? I made the eample the other day of some type à la ValueStringBuilder, returning a span with the content written so far. Even if that buffer was wrapping an array in the heap and therefore would be safe to escape, it would be nice to be able to mark that property as unscoped so that callers would not be able to escape that span outside of the builder scope, as the builder would be disposed when out of scope (say, with a using declaration), and that disposal logic would cause the buffer to be returned to the array pool. Without that keyword, there would be no way to prevent consumers from misusing that and incorrectly escaping that span.

Reply:
Sure. For instance, in the Store, we built a custom ETW provider, which allows users to pass in an interpolated string expression and then have it be null-terminated and passed directly to the native API without having to allocate an intermediary string (we're using this all over the place for all of our ETW logging). The logging APIs exposed to callers look look something like this (simplified):

```csharp
void ILogger.LogInformation(ref ILogger.InterpolatedStringHandler handler)
{
    using var context = handler.DangerousGetContext();

    TheNativePInvokeEtwAPI(context.Text, ...);
}
```

Now, the `ILogger.InterpolatedStringHandler` is essentially very similar to `DefaultInterpolatedStringHandler`, except it exposes that `DangerousGetContext()` API which looks like this, conceptually (this is currently implemented with unsafe hacks):

```csharp
ref struct InterpolatedStringHandler
{
    public Context DangerousGetContext()
    {
        return new(ref this);
    }

    public readonly ref struct Context
    {
        private readonly ref InterpolatedStringHandler _handler;

        internal Context(ref InterpolatedStringHandler handler)
        {
            _handler = handler;
            Text = handler.AppendNullTerminatorAndGetBuffer();
        }

        public ReadOnlySpan<char> Text { get; } // This should be unscoped

        public void Dispose()
        {
            _handler.Clear();
        }
    }
}
```

Essentially, `Context` is the only way to access the raw span, and it ensures the null-terminator is added before you get it. When that's disposed, the parent handler is also disposed so its buffer is returned. Now, there's two caveats:
- The `Context` value returned by `DangerousGetContext` escapes `this`, so it should be `unscoped`. This would be the classic "we're escaping `ref this`" where `unscoped` was meant for.
- The `Text` span exposed by `Context` **also** cannot escape `Context`, as its buffer is owned by `Context` (and transitively, by its parent handler). As such, you'd want to also annotate that `Text` property with `unscoped`, to ensure that its lifetime cannot escape that of its parent `Context`, even though it's not technically escaping a reference to it.

The latter is what I meant for `unscoping` also being very useful for expressing some sort of "ownership" semantics, even in cases where the issue is not strictly speaking an actual ref to an instance escaping out of scope.

Does that make sense?

Reply:
Clarified on Discord, `unscoped` would not support this specific scenario anyway.
Thanks @jaredpar for bearing with me and explaining this in detail 😄
Resolving this.